prow/plugins/projectmanager/projectmanager.go

```diff
@@ -0,0 +1,200 @@
+)
+
+// TODO Create a new handler for issues, look in hook/server.go
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns if they match given criteria.",
+		Config: map[string]string{"project_manager": `orgsRepos: 
+                             kubernetes/kubernetes:
+                               projects:
+                                 test_project:
+                                   columns:
+                                     - name: triage
+                                     state: open
+                                     org: kubernetes
+                                     labels: 
+                                     - area/conformance
+                                   - id: 0
+                                      name: need-attention
+                                      state: open
+                                      org: kubernetes
+                                      labels: 
+                                      - area/apiserver
+                                 test_project2:
+                                   columns:
+                                   - id: 0
+                                      name: triage
+                                      state: open
+                                      org: kubernetes
+                                      labels:
+                                      - sig/testing
+                                   - id: 0
+                                      name: need-attention
+                                      state: open
+                                      org: kubernetes
+                                      labels: 
+                                      - sig/networking`,
+		},
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectManager, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetRepoProjects(owner, repo string) ([]github.Project, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectManager plugins.ProjectManager, log *logrus.Entry, pe github.PullRequestEvent) error {
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
```

Comment:
What happens if the PR was already added to the project and a label is added or the PR is reopened? It seems like this will create a duplicate column card (or fail trying if that is disallowed). This is particularly problematic for the "label added" case which is very common.

Similarly, what happens if a PR stops meeting the criteria? (state change, label change, etc.) Currently this code doesn't attempt to handle that case so project cards are never removed.

Reply:
Currently we cannot add project card to the project if it already exist. Duplication is checked by Git API.  I modifies the fakegithub client to behave similarly
If a label is removed we need to handle the case to remove the project card, this is stated in the TODO items at the top of the module.

Reply:
> Duplication is checked by Git API

IIUC every time a label is added we'll try to add all the column cards. For each of the column cards we waste an API token trying to add the card again when it already exists. Since this can happen for every label event this could be a significant waste, right? Can we cost efficiently find all the column cards for the PR and deduplicate ourselves or would that require the same number of API requests?

Reply:
I think I understand the concern,  git project allows users to add note cards on issue/Prs from any org and repo. This allows projects to track external dependencies.
I wish the PR/Issue api is enhanced to provide info on all projects it belongs to (like how the UI shows in the right side column)
Currently, I believe, the cost of failing with 'Card already exists' is cheaper than maintaining a preloaded cache and maintaining refresh.  This can be improved later too.